### ISSUE TITLE
Subprocesses incorrectly overlapping in some scenarios

### ISSUE THREAD

**Issue by bengallienne** [bpmn-io/bpmn-js#1102]
Hi,

When using the system and having maps that include sub processes, we have noticed that occasionally when you expand a subprocess, other sub processes will overlap it when they should not (see linked video below).

[Example video](https://codewebdevelopment.com/subprocess_overlap.mov)

[example_overlap.bpmn.zip](https://github.com/bpmn-io/bpmn-js/files/3324104/example_overlap.bpmn.zip)

1. Open the attached bpmn file in the online demo (you need to unzip it first because github does not allow .bpmn files)
2. expand the subprocess named ‘subprocess 1’ 
3. see that the subprocess named ‘overlap’ appears on top of it (you can consistently recreate this by expanding and collapsing ‘subprocess 1’, WITHOUT moving it.)
4. Move 'subprocess 1' slightly by dragging, then expand again, the correct order is shown, and ‘overlap’ is hidden
5. If you move ‘overlap’ after that, then expand subprocess 1 again, the issue returns

Expected Behavior

The expanded subprocess should always sit over the top of any overlapping elements, including other subprocesses.

Environment

Please complete the following information:

 - Browser: Chrome latest
 - OS: Mac Os latest
 - Library version: Latest, the version used for your demo

**Comment by pinussilvestrus** [comment]
Bug still located on library version `v3.4.0`. So it's not a regression of the current release.

**Comment by rajgoel** [comment]
@barmac Here is an idea how the modeller could automatically make space when needed:

Let `(oldWidth, oldHeight)` denote the width and height of the collapsed supbprocess and `(oldLeft, oldRight, oldTop, oldBottom)` denote the bounding box of the collapsed subprocess. Let `(newWidth, newHeight)` denote the size of the expanded subprocess `(newLeft, newRight, newTop, newBottom)` denote the unknown bounding box of the expanded subprocess. 

We set 
- `newLeft = oldLeft - ( newWidth - oldWith)/2`
- `newRight = oldRight + ( newWidth - oldWith)/2`
- `newTop = oldTop - ( newHeight - oldHeight)/2`
- `newBottom = oldBottom - ( newHeight - oldHeight)/2`

Let `maxRight` denote the largest right hand side of all other elements left of the collapsed supbprocess.
Let `minLeft` denote the smallest left hand side of all other elements right of the collapsed supbprocess.
Let `maxBottom` denote the largest bottom value of all other elements above the collapsed supbprocess.
Let `minTop` denote the smallest top value of all other elements below the collapsed supbprocess.

- If `maxRight + margin > newLeft` move all other elements left of the collapsed supbprocess left by ` (maxRight + margin) - newLeft`
- If `minLeft - margin < newRight` move all other elements right of the collapsed supbprocess right by `newRight -  (minLeft - margin)`
- If `maxBottom + margin > newTop` move all other elements above the collapsed supbprocess up by ` (maxBottom + margin) - newTop`
- If `minTop - margin < newBottom` move all other elements right of the collapsed supbprocess down by `newBottom -  (minTop - margin)`

Here `margin` is some suitable minimum distance between elements.

The effect should be the same as using the make space tool manually in the four directions.

**Comment by barmac** [comment]
Thanks for sharing your algorithm. I'd love to see this in action some day.

**Comment by nikku** [comment]
@rajgoel I'm not as confident as @barmac on the "making space" algorithm.

I feel it is a recursive layouting problem that could be hard to tackle. Let us for example consider the following diagram; what would be a good layout once `choose recipe` expands to the approximate size of the parent participant, or more?

![image](https://user-images.githubusercontent.com/58601/226757438-f33e3a66-d0cb-4e68-b450-f4dea5afc3b0.png)

**Comment by rajgoel** [comment]
@nikku  You are right I didn't think of collapsed subprocesses nested within expanded subprocesses or participant pools.

However, the general idea could stay the same. The only difference is that one needs to also consider the corners of the parent when determining `maxRight`, `minLeft`, `maxBottom`, `minTop` and move the respective sides of the parent if necessary.

Overall, the rearrangement would look somehow like this manual replication of the approach:
![Screencast 2023-03-22 07 44 42](https://user-images.githubusercontent.com/4448455/226825085-5c6ffba9-6376-4853-9cd6-cce1a2b786b1.gif)
